python/ray/tune/trainable.py

```diff
@@ -227,6 +231,38 @@ def __init__(self, config=None, logger_creator=None):
             self._result_logger = UnifiedLogger(
                 self.config, self._logdir, loggers=None)
 
+        self._stdout_context = self._stdout_fp = self._stdout_stream = None
+        self._stderr_context = self._stderr_fp = self._stderr_stream = None
+        self._stderr_logging_handler = None
+
+        stdout_file = self.config.pop(STDOUT_FILE, None)
+        stderr_file = self.config.pop(STDERR_FILE, None)
+
+        if stdout_file:
+            stdout_path = os.path.expanduser(
+                os.path.join(self._logdir, stdout_file))
+            self._stdout_fp = open(stdout_path, "a+")
+            self._stdout_stream = Tee(sys.stdout, self._stdout_fp)
+            self._stdout_context = redirect_stdout(self._stdout_stream)
+            self._stdout_context.__enter__()
+
+        if stderr_file:
+            stderr_path = os.path.expanduser(
+                os.path.join(self._logdir, stderr_file))
+            self._stderr_fp = open(stderr_path, "a+")
+            self._stderr_stream = Tee(sys.stderr, self._stderr_fp)
+            self._stderr_context = redirect_stderr(self._stderr_stream)
+            self._stderr_context.__enter__()
+
+            # Add logging handler to root ray logger
+            formatter = logging.Formatter("[%(levelname)s %(asctime)s] "
+                                          "%(filename)s: %(lineno)d  "
+                                          "%(message)s")
+            self._stderr_logging_handler = logging.StreamHandler(
+                self._stderr_fp)
+            self._stderr_logging_handler.setFormatter(formatter)
+            ray.logger.addHandler(self._stderr_logging_handler)
```

Comment:
question - what does this do? I was under impression that ray already manages logging/formatting automatically?

Reply:
Pythons logging prints to stderr, but it does not update the output stream when we redirect stderr to a file/Tee object. So in order to have the logs turn up in the stderr file, we have to add a new handler. The default handler only logs the loglevel, thus we need to add a formatter here, too.

Another possibility would be to leave adding a logging handler to the user, who can then decide where to store the logs themselves.